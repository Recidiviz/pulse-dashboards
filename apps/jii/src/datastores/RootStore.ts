// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { captureException } from "@sentry/react";
import { makeObservable, onReactionError } from "mobx";

import { isOfflineMode, isTestEnv } from "~client-env-utils";
import { FlowMethod } from "~hydration-utils";

import { ApiClient } from "../apis/data/ApiClient";
import { DataAPI } from "../apis/data/interface";
import { OfflineAPIClient } from "../apis/data/OfflineAPIClient";
import { StateCode } from "../configs/types";
import { ResidentsStore } from "./ResidentsStore";
import { UiStore } from "./UiStore";
import { UserStore } from "./UserStore";

// global error handling for Mobx reactions
onReactionError((error) => {
  captureException(error);
});

export class RootStore {
  /**
   * Will be an offline client in offline and test modes,
   * and an external API client otherwise.
   */
  apiClient: DataAPI;

  residentsStore?: ResidentsStore;

  userStore: UserStore;

  uiStore: UiStore;

  // for convenience, this is a constant while we only have one state onboarded
  readonly stateCode: StateCode = "US_ME";

  constructor() {
    makeObservable(this, {
      populateResidentsStore: true,
      residentsStore: true,
    });

    this.userStore = new UserStore(this);

    this.uiStore = new UiStore();

    this.apiClient = this.createApiClient();
  }

  /**
   * Constructs and returns an appropriate API client based on the environment
   * (e.g. one that returns fixture data in offline mode). Not intended to be
   * MobX-reactive, since it creates a brand-new instance and discards any state
   * the previous instance might have held.
   */
  private createApiClient(): DataAPI {
    if (isOfflineMode() || isTestEnv()) {
      return new OfflineAPIClient(this);
    } else {
      // eslint-disable-next-line @typescript-eslint/no-this-alias
      const store = this;
      const externals = {
        // these are getters rather than static properties so that the API client
        // can can observe changes to these values rather than getting a one-time snapshot
        get stateCode() {
          return store.stateCode;
        },
        get authClient() {
          return store.userStore.authClient;
        },
        get config() {
          return store.residentsStore?.config;
        },
      };

      return new ApiClient(externals);
    }
  }

  *populateResidentsStore(): FlowMethod<DataAPI["residentsConfig"], void> {
    const config = yield this.apiClient.residentsConfig();

    this.residentsStore = new ResidentsStore(this, config);
  }
}
