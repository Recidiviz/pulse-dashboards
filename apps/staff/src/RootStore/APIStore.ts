// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import dedent from "dedent";

import UserStore from "./UserStore";

interface RequestProps {
  path: string;
  method: "GET" | "POST" | "PATCH" | "DELETE";
  body?: Record<string, unknown>;
  retrying?: boolean;
}

type ExternalSMSMessageRequest = {
  recipientExternalId: string;
  recipientPhoneNumber: string;
  senderId: string;
  message: string;
  userHash: string;
};

export class APIStore {
  userStore: UserStore;

  constructor(userStore: UserStore) {
    this.userStore = userStore;
  }

  async request({
    path,
    method,
    body,
    retrying = false,
  }: RequestProps): Promise<any> {
    if (!this.userStore.getToken) {
      return Promise.reject();
    }

    const token = await this.userStore.getToken();

    const headers: HeadersInit = {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    };

    const response = await fetch(path, {
      body: method !== "GET" ? JSON.stringify(body) : null,
      method,
      headers,
    });

    const json = await response.json();

    if (!response.ok) {
      throw new Error(
        dedent`API request failed.
            Status: ${response.status} - ${response.statusText}
            Errors: ${JSON.stringify(json.errors ?? json)}`,
      );
    }
    return json;
  }

  async get(path: string, body: Record<string, string> = {}): Promise<any> {
    return this.request({ path, body, method: "GET" });
  }

  async post(path: string, body: Record<string, unknown> = {}): Promise<any> {
    return this.request({ path, body, method: "POST" });
  }

  async patch(path: string, body: Record<string, unknown> = {}): Promise<any> {
    return this.request({ path, body, method: "PATCH" });
  }

  async postExternalSMSMessage(body: ExternalSMSMessageRequest): Promise<any> {
    const stateCode = this.userStore.isRecidivizUser
      ? this.userStore.rootStore?.currentTenantId
      : this.userStore.stateCode;
    return this.post(
      `${import.meta.env.VITE_NEW_BACKEND_API_URL}/workflows/external_request/${stateCode}/enqueue_sms_request`,
      body,
    );
  }
}
