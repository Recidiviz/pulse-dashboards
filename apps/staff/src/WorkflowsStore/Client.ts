// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import dedent from "dedent";
import { deleteField, FieldValue, serverTimestamp } from "firebase/firestore";
import { mapValues, toUpper } from "lodash";
import { action, makeObservable, override } from "mobx";
import { format as formatPhone } from "phone-fns";
import { toast } from "react-hot-toast";

import {
  ClientEmployer,
  ClientRecord,
  congratulationsMilestoneTypes,
  Milestone,
  MilestoneType,
  profileMilestoneTypes,
  SpecialConditionCode,
} from "~datatypes";

import { workflowsUrl } from "../core/views";
import {
  DeclineReason,
  MilestonesMessage,
  PortionServedDates,
  TextMessageStatus,
  TextMessageStatuses,
} from "../FirestoreStore/types";
import type { RootStore } from "../RootStore";
import { TENANT_CONFIGS } from "../tenants";
import { JusticeInvolvedPersonBase } from "./JusticeInvolvedPersonBase";
import { MilestonesMessageUpdateSubscription } from "./subscriptions/MilestonesMessageUpdateSubscription";
import { SupervisionTaskInterface } from "./Task/types";
import { UsIdSupervisionTasks } from "./Task/UsIdSupervisionTasks";
import { UsTxSupervisionTasks } from "./Task/UsTxSupervisionTasks";
import { JusticeInvolvedPerson } from "./types";
import {
  clearPhoneNumberFormatting,
  formatSupervisionType,
  fractionalDateBetweenTwoDates,
  OTHER_KEY,
} from "./utils";

export const UNKNOWN = "Unknown" as const;

type ClientDetailsCopy = {
  supervisionStartDate?: {
    tooltip: string;
  };
};

export const CLIENT_DETAILS_COPY: Record<string, ClientDetailsCopy> = {
  US_ND: {
    supervisionStartDate: {
      tooltip:
        "This date is included in the filled form as the date that the Defendant was sentenced; in rare cases, the sentencing date and supervision start date may differ. Double check the date in the form to ensure that the sentencing date and supervision start date match in this case.",
    },
  },
};

export type TaskFactory<PersonType extends JusticeInvolvedPerson> = (
  person: PersonType,
) => SupervisionTaskInterface | undefined;

const createClientSupervisionTasks: TaskFactory<Client> = (
  person,
): SupervisionTaskInterface | undefined => {
  const { allowSupervisionTasks } = person.rootStore.workflowsStore;
  if (!(person instanceof Client && allowSupervisionTasks)) return;

  if (person.rootStore.currentTenantId === "US_TX") {
    return new UsTxSupervisionTasks(person);
  }

  if (person.rootStore.currentTenantId === "US_ID") {
    return new UsIdSupervisionTasks(person);
  }
};

export function isClient(
  person: Client | JusticeInvolvedPerson,
): person is Client {
  return person instanceof Client;
}

// TODO(#1735): the real type should be cleaner than this
export type ParsedSpecialCondition = {
  // eslint-disable-next-line camelcase
  note_update_date: string;
  // eslint-disable-next-line camelcase
  conditions_on_date: string | null;
};

export type ParsedSpecialConditionOrString =
  | NonNullable<ParsedSpecialCondition>
  | string;

function filteredMilestoneTypes(
  milestones: Milestone[] | undefined,
  types: readonly MilestoneType[],
) {
  return (milestones ?? []).filter(({ type }) => types.includes(type));
}

/**
 * Represents a person on supervision
 */
export class Client extends JusticeInvolvedPersonBase<ClientRecord> {
  supervisionLevelStart?: Date;

  address?: string;

  private rawPhoneNumber?: string;

  expirationDate?: Date;

  supervisionStartDate?: Date;

  currentBalance?: number;

  lastPaymentAmount?: number;

  lastPaymentDate?: Date;

  probationSpecialConditions?: string[];

  paroleSpecialConditions?: SpecialConditionCode[];

  currentEmployers?: ClientEmployer[];

  tenantMilestones?: Milestone[];

  emailAddress?: string;

  milestonesMessageUpdatesSubscription?: MilestonesMessageUpdateSubscription<MilestonesMessage>;

  constructor(record: ClientRecord, rootStore: RootStore) {
    super(record, rootStore, createClientSupervisionTasks);

    if (
      rootStore.currentTenantId &&
      TENANT_CONFIGS[
        rootStore.currentTenantId
      ]?.navigation?.workflows?.includes("milestones")
    ) {
      this.milestonesMessageUpdatesSubscription =
        new MilestonesMessageUpdateSubscription(
          this.rootStore.firestoreStore,
          record.recordId,
        );
    }

    makeObservable(this, {
      supervisionLevel: true,
      updateRecord: override,
      milestoneMessagesUpdates: true,
      milestonesPhoneNumber: true,
      milestonesPendingMessage: true,
      milestonesMessageStatus: true,
      updateMilestonesPhoneNumber: action,
      updateMilestonesTextMessage: action,
    });
    this.updateRecord(record);
  }

  updateRecord(record: ClientRecord): void {
    super.updateRecord(record);
    this.supervisionLevelStart = record.supervisionLevelStart;
    this.address = record.address;
    this.rawPhoneNumber = record.phoneNumber;
    this.expirationDate = record.expirationDate;
    this.currentBalance = record.currentBalance;
    this.lastPaymentDate = record.lastPaymentDate;
    this.lastPaymentAmount = record.lastPaymentAmount;
    this.probationSpecialConditions = record.specialConditions;
    this.paroleSpecialConditions = record.boardConditions ?? [];
    this.supervisionStartDate = record.supervisionStartDate;
    this.currentEmployers = record.currentEmployers;
    this.emailAddress = record.emailAddress;

    const tenantMilestoneTypes =
      this.rootStore.currentTenantId &&
      TENANT_CONFIGS[this.rootStore.currentTenantId]?.milestoneTypes;
    this.tenantMilestones = tenantMilestoneTypes
      ? filteredMilestoneTypes(record.milestones, tenantMilestoneTypes)
      : record.milestones;
  }

  get profileUrl(): string {
    return workflowsUrl("clientProfile", {
      justiceInvolvedPersonId: this.pseudonymizedId,
    });
  }

  get assignedStaffId(): string {
    return this.record.officerId;
  }

  get supervisionType(): string {
    const supervisionType = this.record.supervisionType;

    return !supervisionType || supervisionType === "INTERNAL_UNKNOWN"
      ? "Unknown"
      : formatSupervisionType(supervisionType);
  }

  get supervisionLevel(): string {
    return this.rootStore.workflowsStore.formatSupervisionLevel(
      this.record.supervisionLevel,
    );
  }

  get phoneNumber(): string | undefined {
    return this.rawPhoneNumber
      ? formatPhone("(NNN) NNN-NNNN", this.rawPhoneNumber)
      : undefined;
  }

  get detailsCopy(): ClientDetailsCopy {
    return CLIENT_DETAILS_COPY[this.stateCode];
  }

  get district() {
    // For some reason typescript doesn't like super.district
    // eslint-disable-next-line dot-notation
    return this.record.district ?? super["district"];
  }

  get portionServedDates(): PortionServedDates {
    const startDate = this.record.supervisionStartDate;
    const endDate = this.record.expirationDate;

    const opportunityDates: PortionServedDates = [];

    const opportunities = Object.values(this.opportunities).flat();

    opportunities.forEach((opp) => {
      if ("portionServedRequirement" in opp) {
        if (
          opp.portionServedRequirement &&
          opp.portionServedRequirement.includes("1/2")
        ) {
          opportunityDates.push({
            heading: "Half Time",
            date: fractionalDateBetweenTwoDates(startDate, endDate, 0.5),
          });
        }
      }
    });

    return opportunityDates;
  }

  get formattedProbationSpecialConditions(): ParsedSpecialConditionOrString[] {
    // we will flatten the nested lists of conditions into this
    const conditionsToDisplay: ParsedSpecialConditionOrString[] = [];

    // TODO(#1735): after data/ETL change we should expect structured data
    // rather than a JSON-ish string
    this.probationSpecialConditions?.forEach((conditionsJson) => {
      try {
        const conditionsForSentence: {
          // eslint-disable-next-line camelcase
          note_update_date: string;
          // eslint-disable-next-line camelcase
          conditions_on_date: string | null;
        }[] = JSON.parse(
          // the specialConditions strings are almost valid JSON,
          // except they may include NULL instead of null as a value;
          // work around this by converting to lowercase
          conditionsJson.toLowerCase(),
        );

        conditionsForSentence.forEach(
          // eslint-disable-next-line camelcase
          ({ note_update_date, conditions_on_date }) => {
            // don't display nulls
            // eslint-disable-next-line camelcase
            if (!conditions_on_date) return;

            // note that we have to convert the actual values back to uppercase
            // to display them properly
            conditionsToDisplay.push(
              // eslint-disable-next-line camelcase
              mapValues({ note_update_date, conditions_on_date }, toUpper),
            );
          },
        );
      } catch (e) {
        // if we couldn't hack our way to valid JSON,
        // display the whole ugly string so there's no data loss
        conditionsToDisplay.push(conditionsJson);
      }
    });
    return conditionsToDisplay;
  }

  get systemConfig() {
    return this.rootStore.workflowsStore.systemConfigFor("SUPERVISION");
  }

  get profileMilestones(): Milestone[] {
    return filteredMilestoneTypes(this.tenantMilestones, profileMilestoneTypes);
  }

  get congratulationsMilestones(): Milestone[] {
    return filteredMilestoneTypes(
      this.tenantMilestones,
      congratulationsMilestoneTypes,
    );
  }

  get milestoneMessagesUpdates(): MilestonesMessage | undefined {
    return this.milestonesMessageUpdatesSubscription?.data;
  }

  get milestonesMessageStatus(): TextMessageStatus | undefined {
    return this.milestoneMessagesUpdates?.status;
  }

  get milestonesPendingMessage(): string | undefined {
    return this.milestoneMessagesUpdates?.pendingMessage;
  }

  get milestonesMessageErrors(): string[] | undefined {
    return this.milestoneMessagesUpdates?.errors;
  }

  get milestonesFullTextMessage(): string | undefined {
    return this.milestoneMessagesUpdates?.message;
  }

  get milestonesDeclinedReasons(): MilestonesMessage["declinedReasons"] {
    return this.milestoneMessagesUpdates?.declinedReasons;
  }

  get milestonesMessageUpdateLog(): MilestonesMessage["updated"] | undefined {
    return this.milestoneMessagesUpdates?.updated;
  }

  get milestonesPhoneNumber(): string | undefined {
    const userEnteredPhoneNumber = this.milestoneMessagesUpdates?.recipient;
    return userEnteredPhoneNumber?.length
      ? userEnteredPhoneNumber
      : this.rawPhoneNumber;
  }

  milestonesPhoneNumberDoesNotMatchClient(enteredPhoneNumber: string): boolean {
    if (!enteredPhoneNumber || !this.rawPhoneNumber) return false;
    return (
      clearPhoneNumberFormatting(enteredPhoneNumber) !== this.rawPhoneNumber
    );
  }

  async updateMilestonesDeclineReasons(
    reasons: DeclineReason[],
    otherReason?: string,
  ): Promise<void> {
    const otherReasonField = reasons.includes(OTHER_KEY)
      ? { otherReason }
      : { otherReason: deleteField() };

    const updated = {
      by: this.rootStore.workflowsStore.currentUserEmail,
      date: serverTimestamp(),
    };
    await this.rootStore.firestoreStore.updateMilestonesMessages(
      this.recordId,
      {
        updated,
        status: TextMessageStatuses.DECLINED,
        declinedReasons: {
          ...(reasons.length ? { reasons } : {}),
          ...otherReasonField,
        },
      },
    );
  }

  async undoMilestonesDeclined(): Promise<void> {
    const updated = {
      by: this.rootStore.workflowsStore.currentUserEmail,
      date: serverTimestamp(),
    };
    await this.rootStore.firestoreStore.updateMilestonesMessages(
      this.recordId,
      {
        updated,
        status: TextMessageStatuses.PENDING,
        declinedReasons: deleteField(),
      },
    );
  }

  async updateMilestonesPhoneNumber(
    phoneNumber: string,
    deletePhoneNumber = false,
  ): Promise<void> {
    const updated = {
      by: this.rootStore.workflowsStore.currentUserEmail,
      date: serverTimestamp(),
    };
    await this.rootStore.firestoreStore.updateMilestonesMessages(
      this.recordId,
      {
        updated,
        status: TextMessageStatuses.PENDING,
        stateCode: this.stateCode,
        recipient: deletePhoneNumber ? deleteField() : phoneNumber,
        userHash: this.rootStore.workflowsStore.currentUserHash,
      },
    );
  }

  async updateMilestonesTextMessage(
    additionalMessage?: string,
    deleteAdditionalMessage = false,
  ): Promise<void> {
    let pendingMessage: Partial<Record<"pendingMessage", string | FieldValue>>;
    if (deleteAdditionalMessage) {
      pendingMessage = { pendingMessage: deleteField() };
    } else if (additionalMessage && additionalMessage !== "") {
      pendingMessage = {
        pendingMessage: additionalMessage,
      };
    } else {
      pendingMessage = {};
    }

    const updated = {
      by: this.rootStore.workflowsStore.currentUserEmail,
      date: serverTimestamp(),
    };
    await this.rootStore.firestoreStore.updateMilestonesMessages(
      this.recordId,
      {
        updated,
        userHash: this.rootStore.workflowsStore.currentUserHash,
        status: TextMessageStatuses.PENDING,
        ...pendingMessage,
        stateCode: this.stateCode,
        message: dedent`
            ${this.defaultMilestonesMessage}

            ${additionalMessage || ""}
          `,
      },
    );
  }

  async updateMilestonesStatus(status: TextMessageStatus) {
    await this.rootStore.firestoreStore.updateMilestonesMessages(
      this.recordId,
      {
        updated: {
          by: this.rootStore.workflowsStore.currentUserEmail,
          date: serverTimestamp(),
        },
        status,
        userHash: this.rootStore.workflowsStore.currentUserHash,
      },
    );
  }

  async sendMilestonesMessage(): Promise<void> {
    if (!this.milestonesFullTextMessage || !this.milestonesPhoneNumber) return;
    try {
      await this.rootStore.firestoreStore.updateMilestonesMessages(
        this.recordId,
        {
          updated: {
            by: this.rootStore.workflowsStore.currentUserEmail,
            date: serverTimestamp(),
          },
          status: TextMessageStatuses.IN_PROGRESS,
          userHash: this.rootStore.workflowsStore.currentUserHash,
        },
      );
      await this.rootStore.apiStore.postExternalSMSMessage({
        message: this.milestonesFullTextMessage,
        recipientExternalId: this.externalId,
        recipientPhoneNumber: this.milestonesPhoneNumber,
        senderId: this.rootStore.workflowsStore.currentUserEmail ?? "Unknown",
        userHash: this.rootStore.workflowsStore.currentUserHash,
      });
    } catch (e) {
      await this.rootStore.firestoreStore.updateMilestonesMessages(
        this.recordId,
        {
          updated: {
            by: this.rootStore.workflowsStore.currentUserEmail,
            date: serverTimestamp(),
          },
          status: TextMessageStatuses.PENDING,
          userHash: this.rootStore.workflowsStore.currentUserHash,
        },
      );
      toast.error(
        "We couldn't send your message. Please wait a moment and try again.",
      );
      throw e;
    }
  }

  get defaultMilestonesMessage(): string {
    const {
      rootStore: {
        userStore: { userSurname },
      },
      congratulationsMilestones,
    } = this;

    const clientName = this.preferredName ?? this.fullName.givenNames;
    const salutation = clientName ? `Hey ${clientName}!` : `Hey!`;

    return dedent`
    Message from Agent ${userSurname} at ${
      !this.rootStore.currentTenantId
        ? "DOC"
        : TENANT_CONFIGS[this.rootStore.currentTenantId].DOCName
    }:

    ${salutation} Congratulations on reaching ${
      congratulationsMilestones.length > 1
        ? "these milestones"
        : "this milestone"
    }:

    ${congratulationsMilestones.map((m) => `- ${m.text}`).join("\n")}
  `;
  }
}
