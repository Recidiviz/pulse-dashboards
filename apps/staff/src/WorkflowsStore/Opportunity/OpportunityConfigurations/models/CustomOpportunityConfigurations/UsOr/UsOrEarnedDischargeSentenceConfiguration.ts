// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { OpportunityRequirement } from "../../../../types";
import { ApiOpportunityConfiguration } from "../../ApiOpportunityConfigurationImpl";

export class UsOrEarnedDischargeSentenceConfiguration extends ApiOpportunityConfiguration {
  get omsCriteriaHeader() {
    return "Eligibility Criteria Requirements Verified via DOC400";
  }

  get nonOMSCriteriaHeader() {
    return "Additional Eligibility Requirements Manually Verified ";
  }

  get nonOMSCriteria(): OpportunityRequirement[] {
    return [
      {
        text: "No disqualifying sentence enhancements not entered in the DOC400",
        tooltip:
          "Is not on supervision for a sentencing enhancement imposed under the provisions of ORS 161.610, 161.725, 161.735, 164.061, 475.907, 475.925, or 475.930; as well as ORS 137.635 for Probation Burglary I ",
      },
      {
        text: "No court violations in the past 6 months",
        tooltip:
          "Has not been found in violation by the court in the immediate 6 months prior to review",
      },
      {
        text: "Not convicted of a crime that occurred while on supervision for the case under review (not found in DOC400/CIS)",
        tooltip:
          "Has not been convicted of a crime (felony or misdemeanor) that occurred while on supervision for the case under review",
      },
      {
        text: "Has fully paid any restitution and compensatory fine or is current on payment plan",
        tooltip:
          "Has either fully paid any restitution and compensatory fine ordered by the court, or established a payment schedule through the court or appropriate supervising authority consistent with ORS 137.106, and is current in their payment obligations",
      },
      {
        text: "Has completed any specialty court programs or treatment programs",
        tooltip:
          "Has completed any specialty court program and treatment programs with set durations or timeframes, and has consistently participated in any ongoing treatment programs",
      },
      {
        text: "In compliance with supervision conditions and case plan",
        tooltip:
          "Is in compliance with conditions of supervision and any applicable supervision case plan",
      },
    ];
  }
}
