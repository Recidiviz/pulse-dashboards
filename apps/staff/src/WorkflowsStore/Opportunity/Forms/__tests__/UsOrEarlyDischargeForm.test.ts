// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { parseISO } from "date-fns";
import { configure } from "mobx";
import tk from "timekeeper";

import { RootStore } from "../../../../RootStore";
import { Client } from "../../../Client";
import { UsOrEarnedDischargeOpportunity } from "../../UsOr";
import { UsOrEarnedDischargeForm } from "../UsOrEarnedDischargeForm";

let form: UsOrEarnedDischargeForm;
let opp: (typeof form)["opportunity"];
let personRecord: (typeof opp)["person"]["record"];
let oppRecord: (typeof opp)["record"] & object;
let staffRecord: (typeof opp)["person"]["assignedStaff"];

function createTestUnit() {
  const rootStore = new RootStore();
  personRecord = {
    personType: "CLIENT",
    stateCode: "US_OZ",
    recordId: "US_OZ1",
    pseudonymizedId: "pseudo1",
    displayId: "d1",
    personExternalId: "pei1",
    personName: { givenNames: "Joe", middleNames: "Quimby", surname: "Test" },
    officerId: "zzz",
    allEligibleOpportunities: ["usOrEarnedDischarge"],
  };
  oppRecord = {
    stateCode: "US_OZ",
    externalId: "pei1",
    metadata: {
      programs: [],
    },
    subOpportunities: [
      {
        eligibleCriteria: {
          eligibleStatute: {},
          noAdministrativeSanction: {},
          noConvictionDuringSentence: {},
          pastHalfCompletionOrSixMonths: {},
        },
        ineligibleCriteria: {},
        id: 123,
        metadata: {
          sentenceId: 123,
          courtCaseNumber: "cc1",
          sentenceSubType: "subtype",
          sentenceImposedDate: parseISO("2020-01-01"),
          sentenceStartDate: parseISO("2020-01-05"),
          sentenceEndDate: parseISO("2022-03-08"),
          sentenceCounty: "COUNTY",
          chargeCounty: "CHARGE COUNTY",
          judgeFullName: "Judge Reinhold",
          sentenceStatute: "STATUTE",
          conditions: [],
        },
      },
    ],
  };
  staffRecord = {
    id: "zzz",
    stateCode: "US_OZ",
    givenNames: "8675309",
    surname: "8675309",
    email: null,
  };
  const person = new Client(personRecord, rootStore);
  opp = new UsOrEarnedDischargeOpportunity(person);
  vi.spyOn(opp, "record", "get").mockImplementation(() => oppRecord);
  vi.spyOn(person, "assignedStaff", "get").mockReturnValue(staffRecord);
  form = opp.form;
}

beforeEach(() => {
  configure({ safeDescriptors: false });
  createTestUnit();
  tk.freeze(new Date("2023-12-12"));
});

afterEach(() => {
  configure({ safeDescriptors: true });
  vi.resetAllMocks();
  tk.reset();
});

describe("prefilledDataTransformer", () => {
  test("basic transformation", () => {
    expect(form.prefilledDataTransformer()).toMatchInlineSnapshot(`
      {
        "clientId": "pei1",
        "countyAddress": "(Your County Address}",
        "countyFax": "(   )",
        "countyName": "(Your County}",
        "countyPhone": "(   )",
        "givenNames": "Joe",
        "middleNames": "Quimby",
        "officerName": "8675309 8675309",
        "sentences": {
          "123": {
            "county": "COUNTY",
            "docket": "cc1",
            "judgeName": "Judge Reinhold",
            "offenses": "STATUTE",
            "sentenceExpirationDate": "Mar 8, 2022",
            "sentenceStartDate": "Jan 5, 2020",
            "sentenceType": "subtype",
          },
        },
        "surname": "Test",
        "todaysDate": "Dec 12, 2023",
      }
    `);
  });

  test("transform multiple sentences", () => {
    const criteria = {
      eligibleCriteria: {
        eligibleStatute: {},
        noAdministrativeSanction: {},
        noConvictionDuringSentence: {},
        pastHalfCompletionOrSixMonths: {},
      },
      ineligibleCriteria: {},
    };
    const multioppRecord = {
      stateCode: "US_OZ",
      externalId: "pei1",
      metadata: {
        programs: [],
      },
      subOpportunities: [
        {
          ...criteria,
          id: 123,
          metadata: {
            sentenceId: 123,
            courtCaseNumber: "cc1",
            sentenceSubType: "subtype",
            sentenceImposedDate: parseISO("2020-01-01"),
            sentenceStartDate: parseISO("2020-01-05"),
            sentenceEndDate: parseISO("2022-03-08"),
            sentenceCounty: "COUNTY",
            chargeCounty: "CHARGE COUNTY",
            judgeFullName: "Judge Reinhold",
            sentenceStatute: "STATUTE",
            conditions: [],
          },
        },
        {
          ...criteria,
          id: 456,
          metadata: {
            sentenceId: 456,
            courtCaseNumber: "cc2",
            sentenceSubType: "subtype2",
            sentenceImposedDate: parseISO("2020-04-04"),
            sentenceStartDate: parseISO("2020-04-05"),
            sentenceEndDate: parseISO("2022-08-08"),
            sentenceCounty: "COUNTY",
            chargeCounty: "CHARGE COUNTY",
            judgeFullName: "Judge Judge",
            sentenceStatute: "STATUTE",
            conditions: [],
          },
        },
        {
          ...criteria,
          id: 789,
          metadata: {
            sentenceId: 789,
            courtCaseNumber: "cc3",
            sentenceSubType: "subtype3",
            sentenceImposedDate: parseISO("2020-02-01"),
            sentenceStartDate: parseISO("2020-07-05"),
            sentenceEndDate: parseISO("2022-06-08"),
            sentenceCounty: "COUNTY",
            chargeCounty: "CHARGE COUNTY 3",
            judgeFullName: "Judge Jury",
            sentenceStatute: "STATUTE",
            conditions: [],
          },
        },
      ],
    };
    vi.spyOn(opp, "record", "get").mockImplementation(() => multioppRecord);
    expect(form.prefilledDataTransformer()).toMatchSnapshot();
  });
});
