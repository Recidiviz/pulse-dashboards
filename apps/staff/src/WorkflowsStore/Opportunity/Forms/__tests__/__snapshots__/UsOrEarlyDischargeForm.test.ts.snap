// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prefilledDataTransformer > transform multiple sentences 1`] = `
{
  "birthdate": "Jan 1, 1990",
  "clientId": "pei1",
  "countyAddress": "(Your County Address}",
  "countyFax": "(   )",
  "countyName": "(Your County}",
  "countyPhone": "(   )",
  "givenNames": "Joe",
  "middleNames": "Quimby",
  "officerName": "8675309 8675309",
  "sentences": {
    "123": {
      "county": "COUNTY",
      "docket": "cc1",
      "judgeName": "Judge Reinhold",
      "offenses": "STATUTE",
      "sentenceExpirationDate": "Mar 8, 2022",
      "sentenceStartDate": "Jan 5, 2020",
      "sentenceType": "subtype",
    },
    "456": {
      "county": "COUNTY",
      "docket": "cc2",
      "judgeName": "Judge Judge",
      "offenses": "STATUTE",
      "sentenceExpirationDate": "Aug 8, 2022",
      "sentenceStartDate": "Apr 5, 2020",
      "sentenceType": "subtype2",
    },
    "789": {
      "county": "COUNTY",
      "docket": "cc3",
      "judgeName": "Judge Jury",
      "offenses": "STATUTE",
      "sentenceExpirationDate": "Jun 8, 2022",
      "sentenceStartDate": "Jul 5, 2020",
      "sentenceType": "subtype3",
    },
  },
  "surname": "Test",
  "todaysDate": "Dec 12, 2023",
}
`;
